pkg/eosclient/eosgrpc/eos_http/eoshttp.go

```diff
@@ -0,0 +1,500 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package eoshttp
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/logger"
+)
+
+// Options to configure the Client.
+type Options struct {
+
+	// HTTP URL of the EOS MGM.
+	// Default is https://eos-example.org
+	BaseURL string
+
+	// Timeout in seconds for connecting to the service
+	ConnectTimeout int
+
+	// Timeout in seconds for sending a request to the service and getting a response
+	// Does not include redirections
+	RWTimeout int
+
+	// Timeout in seconds for performing an operation. Includes every redirection, retry, etc
+	OpTimeout int
+
+	// If the URL is https, then we need to configure this client
+	// with the usual TLS stuff
+	// Defaults are /etc/grid-security/hostcert.pem and /etc/grid-security/hostkey.pem
+	ClientCertFile string
+	ClientKeyFile  string
+
+	// These will override the defaults, which are common system paths hardcoded
+	// in the go x509 implementation (why did they do that?!?!?)
+	// of course /etc/grid-security/certificates is NOT in those defaults!
+	ClientCADirs  string
+	ClientCAFiles string
+}
+
+// We want just one instance of these options in the whole app, as we don't want to
+// instantiate more than once the http client internals. For example to have
+// http keepalive, pools, etc...
+var httpTransport *http.Transport
```

Comment:
Don't keep httpTransport as a global variable. Make a function return it

Reply:
How do I make sure it's just one instance in that case?

Reply:
You can just declare it when you're initializing the HTTP client. I think that's needed only in the New method but I see you're doing that in GETFile and PUTFile as well. Why do we need to initialize it again?

Reply:
The idea was to be able to refresh it if anything goes bad with connection pools or CRLs, and that was the reason for the mutex, to be able to refresh it from a different thread (possibly in comm troubles). Too bad that golang makes a singleton so awkward

Reply:
httpTransport is just a config object which stays constant except at init. Keep it as a member of the Client struct and reuse it

Reply:
no, it's not. httpTransport is a complex object that caches connections. If I keep it in the Client class then it will leak. It has to be unique in the whole app

https://golang.org/pkg/net/http/#Transport

"Transports should be reused instead of created as needed."